### ISSUE TITLE
How to Make Process Asynchronous

### ISSUE THREAD

**Issue by jpgilchrist** [ralphhanna/bpmn-server#6]
With a long running process such as:

![iscout-web-platinum](https://user-images.githubusercontent.com/5065660/88099612-641f5100-cb69-11ea-8025-8e466df7f8bb.png)

What's the best method for making parts or all of it asynchronous? In camunda, you could mark a node as asynchronous before or after, which would result in the execution returning immediately. Thus freeing up resources to continue operating. I see User Tasks and Boundary Events seem to do this in bpmn-server per a few tests I've read through and go into a waiting state. What I'd like to achieve is to hand the process off to a worker thread of sorts and just return the immediate execution instance. Hopefully, I'm describing that well enough. 

I definitely don't want the whole process to finish before `await server.engine.start(name, data)` returns.

**Comment by ralphhanna** [comment]
Very interesting question, I struggled with this for sometime. It has a lot to do with Node.js behavior !!

Let me ask this, what If don't wait for engine.start()  and go ahead with your logic?

To check on status or to get the instance Id, you fire a query using a unique id like case Id

```javascript
  let instance= server.engine.get({ "data.caseId": 5001 });
```

Alternatively, we can add an option to engine.start to return only after creating the instance.
or, as we can have a node marked as asynchronous? not sure what is the syntax? 

In any case, the engine returns when it encounters a wait state as you noted, in (user input, timer, throw messages)

**Comment by jpgilchrist** [comment]
@ralphhanna yeah, that's an adequate suggestion. It does however require a unique (ish) caseId or some key to associate, whereas that really is the id and it'd be nice to have that back from the API itself. I'm in the process of learning BPMN and don't know all the notations yet. Is there a defacto standard notation for marking something as asynchronous? 

Camunda's is `camunda:asyncAfter="true"`, however, that probably means BPMN doesn't actually include this in the spec. We could either support their extension or add one? 

Personally, I think there's the best of both worlds choice: 
* optional parameter to start to return the instance itself, but add a property on the instance that represents the running execution that can be awaited. `let instance = server.engine.start(name, data, nonBlocking); let executionContext = await instance.worker;`
* support camunda:asyncBefore/After or a custom notation

You said, the engine returns when it encounters a wait state such as "user input, timer, throw messages". However, in my example above I do have the intermediate timer event creating a loop. That specific node did not cause the engine to return there, it ran all the way through. Is it supposed to return and wait at that node?

**Comment by jpgilchrist** [comment]
Follow up question: are you using any sort of thread pool to handle the executions or is it running on the main thread? It might be nice to look into something like https://github.com/piscinajs/piscina to run the bulk of the executions, that way when you run it in conjunction with express or something, they don't compete for resources?

**Comment by ralphhanna** [comment]
Here is another way of handling it, using listener 

```js
    const logger = new Logger({ toConsole: false });

    const listener = new EventEmitter();

    listener.on('all', function ({ context, event, }) {
        let msg = '';
        if (context.item)
            msg = ' Item: ' + context.item.elementId;
        console.log('---Event: -->' + event + msg );
    });

    const server = new BPMNServer(configuration, logger, { cron: false });

    // notice no await for next line
    server.engine.start('serviceTask', { v1: 1, v2: 2 }, listener);

```

**Comment by ralphhanna** [comment]
as for returning at timer, there are few examples, see the one about boundary events, it should return while waiting for the timer,
It is possible however, that your timer was not long enough, since it also waits for saving the instance to MongoDB.

I will at your suggestion, thanks

**Comment by jpgilchrist** [comment]
Actually, was just reading more through the code specifically Node.ts `execute` function and you wouldn't necessarily want it to return the same `NODE_ACTION.wait`, but rather you do need it to return the instance and continue processing as it's asynchronous, not necessarily blocking. 

In response, to your above comment, I do see that; I could use the listener to correlate the instance. 

I'll extend the timer and see what happens. Furthermore, if that's the case I may just throw a timer at the very beginning after thee start node and have it enter a wait state like that. Thoughts on that approach?

**Comment by ralphhanna** [comment]
The question is why? what information do you need to know after the execution, why not let it just run without delay and monitor it with listeners?

**Comment by ralphhanna** [comment]
As for the code, what is what is actually happening at the token and execution level,  it returns if all tokens are in wait state.
Token are like threads,  since there can be parallel flows running

**Comment by ralphhanna** [comment]
this is an example of two tokens running in sync:

![image](https://user-images.githubusercontent.com/11893416/88122110-b8d8c100-cb95-11ea-8e83-0c3ec6627824.png)

```js
..token: 0 - end - current: gateway_1 from root {"caseId":6040,"needsRepairs":"Yes","needsCleaning":"Yes"}
..token: 1 - wait - current: task_clean from gateway_1 {"caseId":6040,"needsRepairs":"Yes","needsCleaning":"Yes"}
.token: 2 - wait - current: task_repair from gateway_1 {"caseId":6040,"needsRepairs":"Yes","needsCleaning":"Yes"}
```

**Comment by ralphhanna** [comment]
revised code for the above 
```js
    const logger = new Logger({ toConsole: false });

    const listener = new EventEmitter();

    listener.on('all', function ({ context, event, }) {
        let msg = '';
        if (context.instance.id)
            msg = ' instanceId: ' + context.instance.id;
        if (context.item)
            msg += ' Item: ' + context.item.elementId + " itemId: "+ context.item.id;
        console.log('---Event: -->' + event + msg );
    });

    const server = new BPMNServer(configuration, logger, { cron: false });

    // notice no await for next line
    server.engine.start('serviceTask', { v1: 1, v2: 2 }, listener);
```

**Comment by jpgilchrist** [comment]
That all makes sense. The reason, I've requested it is so that I can integrate it into a system I'm already running. It's a little redundant, but the front end starts a process/queue. It then polls to see if that job is done. Some of the steps in my above diagram can be quite time consuming, so ideally it's run asynchronously. The listener may do the trick though. 

As for the timer, I increased it to 5 Minutes

```xml
    <bpmn:intermediateCatchEvent id="Event_1c0a18y" name="Wait 5 Minutes">
      <bpmn:incoming>Flow_1qh2g3g</bpmn:incoming>
      <bpmn:incoming>PIPELINE_RUNNING_FLOW</bpmn:incoming>
      <bpmn:outgoing>Flow_1gxayaw</bpmn:outgoing>
      <bpmn:timerEventDefinition id="TimerEventDefinition_0pgj7xv">
        <bpmn:timeDuration xsi:type="bpmn:tFormalExpression">PT5M</bpmn:timeDuration>
      </bpmn:timerEventDefinition>
    </bpmn:intermediateCatchEvent>
```

When running the process the following is output:

```
 expression PT5M base dateTue Jul 21 2020 21:05:41 GMT-0400 (Eastern Daylight Time) -> delay of 300 sec 5 min0.08333333333333333 hours 
checkPipeline
```

Note my service function for that check pipeline simply logs "checkPipeline". It definitely doesn't wait a full 5 Minutes.  It does, however, return the ExecutionContext as suggested above before all the checkPipeline loops finish running.

**Comment by jpgilchrist** [comment]
I apologize for the confusing thread of comments here. 

One follow up on the async behavior, when being started from an API request, it'd be nice to have the response come back immediately and the execution runs asynchronously. Instead of waiting to hit a wait state and then return to the client.

**Comment by jpgilchrist** [comment]
Also, my trick for inserting a timer after the start node does the trick quite well with no added code/listeners.

**Comment by ralphhanna** [comment]
reason being for UI access, the user likes to see the results of execution right-away, and most processes are very fast.

I am considering adding an option 
engine.start(...., { async: true }); 

or a separate method : engine.startAsync(...); but in node.js it means having a  Promise

**Comment by jpgilchrist** [comment]
Yeah, I'm sure in 90% of cases in the BPMN space, the executions are fast or are at least fast in between various wait states. However, I doubt I'll be the last to be looking to have more long running/polling loops that depend on an external system. It's not user input, but it is external.

**Comment by jpgilchrist** [comment]
This is good though, definitely making progress on accomplishing my near term goals with this, but definitely want to see it go further! 👍

**Comment by ralphhanna** [comment]
You may also consider a message event to receive input from external source

**Comment by jpgilchrist** [comment]
Sadly, not all external sources send events/messages, but yes I've thought of that too.

**Comment by jpgilchrist** [comment]
LMFAO so I was debugging through the Timer Behaviour and realized I had copied a configuration from your examples with a forceTimersDelay of 1000.... So, that solves that part of this.

**Comment by jpgilchrist** [comment]
Hmm, after removing the configuration options for timers, I'm still noticing that they don't trigger properly. PT5M outputs a wait of 300 seconds: 
```
expression PT5M base dateWed Jul 22 2020 12:59:58 GMT-0400 (Eastern Daylight Time) -> delay of 300 sec 5 min0.08333333333333333 hours 
```
but is running immediately. I'm trying to figure out how all this works; digging deeper.

**Comment by ralphhanna** [comment]
is it a still a problem? if so, just import the process to the demo site and I will debug it for you.

**Comment by ralphhanna** [comment]
Hi James, I have a new release 1.1.7 that improves logging of the timer 


..------timer running ---
--
timer is set at 1595569601.694 - 1970-01-19T11:12:49.601Z
..setting timer for 300 seconds

I have tested it in the demo site, working fine.

please perform npm update to get the latest release, beware that the service call has changes to service(inputValues,context)

take care

**Comment by ralphhanna** [comment]
Here is what I am working on to make `services` asynchronous:
```js 
myService1(input, context) {

    context.response.action = 'wait';
    this.continueWorking(context);
    return;
    // process will go into wait state , system will save instance and return control
}
continueWorking(context) {


    // ... do your work here - takes whatever time...


    context.server.engine.invoke({ id: context.item.id});

}

```
any thoughts?
